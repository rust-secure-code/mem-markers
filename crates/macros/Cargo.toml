[package]
name = "macros"
version = "0.1.0"
authors = ["Ryan Levick <ryan.levick@gmail.com>"]
edition = "2018"

[lib]
proc_macro = true

[[test]]
name = "tests"
path = "tests/diagnostics.rs"

[dependencies]
quote = "1.0"
syn = "1.0"
proc-macro2 = "1.0"

[dev-dependencies]
trybuild = "1.0"
mem-markers = { path = "../.." }
