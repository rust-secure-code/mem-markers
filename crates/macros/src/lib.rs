extern crate proc_macro;

use proc_macro::TokenStream;
use syn::{parse_macro_input, DeriveInput};

mod byte_complete;
mod fixed_layout;
mod utils;

/// Derive `FixedLayout` for a type
#[proc_macro_derive(FixedLayout)]
pub fn fixed_layout_derive(item: TokenStream) -> TokenStream {
    let input = parse_macro_input!(item as DeriveInput);
    fixed_layout::expand(input)
        .unwrap_or_else(|err| err.to_compile_error())
        .into()
}

/// Derive `ByteComplete` for a type
#[proc_macro_derive(ByteComplete)]
pub fn byte_complete_derive(item: TokenStream) -> TokenStream {
    let input = parse_macro_input!(item as DeriveInput);
    byte_complete::expand(input)
        .unwrap_or_else(|err| err.to_compile_error())
        .into()
}
