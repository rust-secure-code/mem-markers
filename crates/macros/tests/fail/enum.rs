use macros::FixedLayout;

#[derive(FixedLayout)]
#[repr(C)]
enum Foo {
    A,
}

fn main() {}
