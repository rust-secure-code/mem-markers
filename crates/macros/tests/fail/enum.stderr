error: FixedLayout on enums is not currently supported
 --> $DIR/enum.rs:3:10
  |
3 | #[derive(FixedLayout)]
  |          ^^^^^^^^^^^
